### ISSUE TITLE
SOSP 25 | DCP: Addressing Input Dynamism In Long-Context  Training via Dynamic Context Parallelism

### ISSUE THREAD

**Issue by woshipapa** [AI-Infra-Team/awesome-papers#116]
# DCP: Addressing Input Dynamism In Long-Context Training via Dynamic Context Parallelism

---

* **Chenyu Jiang**
    * [jchenyu@connect.hku.hk](mailto:jchenyu@connect.hku.hk)
    * [0009-0006-5714-3872](https://orcid.org/0009-0006-5714-3872)
    * The University of Hong Kong
* **Zhenkun Cai**
    * [zkcai@amazon.com](mailto:zkcai@amazon.com)
    * Amazon Web Services
* **Ye Tian**
    * [yetiansh@connect.hku.hk](mailto:yetiansh@connect.hku.hk)
    * The University of Hong Kong
* **Zhen Jia**
    * [zhej@amazon.com](mailto:zhej@amazon.com)
    * Amazon Web Services
* **Yida Wang**
    * [wangyida@amazon.com](mailto:wangyida@amazon.com)
    * Amazon Web Services
* **Chuan Wu**
    * [cwu@cs.hku.hk](mailto:cwu@cs.hku.hk)
    * The University of Hong Kong

[Paper Link](https://arxiv.org/html/2510.10620v1#:~:text=In%20this%20paper%2C%20we%20present%20DCP%2C%20a%20dynamic,fine-grained%20blockwise%20partitioning%20of%20both%20data%20and%20computation.)

---

## Motivation

<img width="757" height="348" alt="Image" src="https://github.com/user-attachments/assets/84c85e29-8af5-41f9-a0d6-3981c1030a7e" />

### 1. Core Bottleneck: The 'Static' Nature of Existing Context Parallelism (CP) Methods
As model context windows expand to 128K or even 2M tokens, Context Parallelism (CP) has become an essential technique. However, existing CP frameworks (e.g., TransformerEngine, LoongTrain) adopt a "static" parallelization configuration. This means they apply the same fixed data partitioning and placement strategy to all training batches.

<img width="802" height="560" alt="Image" src="https://github.com/user-attachments/assets/16936130-7f64-4bcf-b763-288dea7ad2bf" />

### 2. Core Challenge: The 'Dynamic' Nature of Training Data
The paper's core motivation stems from the inability of these static methods to efficiently handle the "input dynamism" prevalent in real-world training data. This dynamism manifests in two primary ways:

* **Variance in Sequence Lengths:** Real-world training sets (e.g., LongAlign) exhibit a highly skewed sequence length distribution, with a vast majority of short sequences and very few long ones (as shown in Fig. 2). Static CP strategies forcibly partition these short sequences across all GPUs, leading to significant redundant communication (as shown in Fig. 5a).

   
<img width="492" height="742" alt="Image" src="https://github.com/user-attachments/assets/5c3a2b06-0010-4609-8999-1503201c7eb0" />

* **Variance in Attention Patterns:** Modern LLM training is no longer limited to the standard Causal Mask. For efficiency or specific scenarios (like ICL and RLHF), a variety of sparse or structured masks have emerged (e.g., $\Lambda$-shaped, Causal blockwise, Shared question in Fig. 6). Existing static CP schemes are optimized specifically for causal masks. When forcibly applied to these new sparse patterns, they result in severe load imbalance and redundant communication (as shown in Fig. 7).

### 3. Ultimate Goal: Build a 'Dynamic and Adaptive' Parallel Framework (DCP)
In summary, the paper's motivation is to resolve the fundamental conflict between static parallelization strategies and dynamic training data. To address this, the paper aims to design a Dynamic Context Parallelism (DCP) framework. The framework's core objective is to achieve fine-grained control, capable of generating a customized parallelization configuration at runtime for each distinct input iteration. This scheme must simultaneously achieve two goals: 1. Minimize communication overhead, and 2. Maintain computational and memory load balance.

---

## Solution

<img width="1129" height="637" alt="Image" src="https://github.com/user-attachments/assets/e4c7064d-a83b-4b56-ad9c-5737fe474402" />

### 1. Block Generator

The foundational step of the DCP solution is to construct a unified, fine-grained representation that decomposes the attention operator into two fundamental units: "Data Blocks" and "Computation Blocks."

* **Data Blocks:** These are the input and output tensors of the attention computation (i.e., $Q, K, V, \text{and } O$). DCP partitions these tensors along the head dimension and the sequence length (L) dimension to generate a series of standardized, smaller data units.
* **Computation Blocks:** These are defined as the most elemental computational units. Each computation block represents a single attention operation between one $Q$ data block ($Q_i$) and one $KV$ data block ($KV_j$).

This representation is key to solving the input dynamism challenge. Its core advantage lies in its intrinsic coupling with the attention mask: a computation block is only generated if the interaction between its corresponding $Q_i$ and $KV_j$ blocks is not masked out. For sparse or structured masks (e.g., a shared question mask), the computation for masked-out interactions is never generated in the first place.

The effects are twofold:
* **Elimination of Redundant Computation:** It innately skips all invalid computations defined by sparse masks.
* **Unification of Heterogeneous Inputs:** It transforms both forms of input dynamism—variable sequence lengths and diverse attention patterns—into a variable, yet standardized, collection of blocks.

### 2. Hypergraph Partitioning for Placement

<img width="783" height="495" alt="Image" src="https://github.com/user-attachments/assets/39f7c408-887f-4cfa-be33-0d10dd54737e" />

**Problem:** Now that we have a collection of "data blocks" and "computation blocks," how do we decide which GPU to "place" them on?

**Objective:** The placement strategy must satisfy two conditions simultaneously:
1.  **Minimize Communication:** $\min \sum s_e (\lambda_e - 1)$
2.  **Maintain Load Balance:** $w(P_i) \preceq [1+\epsilon, 1] \odot w(\mathcal{N})/R$

**DCP's Solution:** DCP models this complex placement problem as a classic **"Balanced Hypergraph Partitioning"** problem.

1.  **Hierarchical Placement:** To prioritize minimizing the most expensive **inter-machine communication**, DCP employs a hierarchical strategy. It first partitions blocks at the "machine" (Node) level, and then performs a second partition for the "devices" (GPUs) within each machine.

2.  **Hypergraph Construction:**
    * **Vertices ($\mathcal{N}$):** All "data blocks" ($\mathcal{I}, \mathcal{O}$) and "computation blocks" ($\mathcal{C}$).
    * **Vertex Weights:** Each vertex is assigned a 2D weight: `[Computation Load, Data Size]`. A computation block has a weight of `[FLOPS, 0]`, while a data block has a weight of `[0, Size]`.
    * **Hyperedges ($\mathcal{E}$):** A single hyperedge $e$ **connects one "data block" $d$ to all its associated "computation blocks" $C_i$** (i.e., all tasks that either consume or produce $d$). The hyperedge's weight $s_e$ is set to the size of that data block.

3.  **Solving:**
    * Through this model, the "total communication volume" is precisely represented by the objective function $\min \sum s_e (\lambda_e - 1)$ (where $\lambda_e$ is the number of partitions a hyperedge spans).
    * "Load balance" is represented as a constraint (i.e., the total vertex weight of each partition $P_i$ must remain close to the average).
    * This is an NP-hard problem, so DCP uses off-the-shelf, efficient heuristic solvers (like PaToH or KaHyPar) to quickly find a near-optimal placement solution.

### 3. Pipelined Scheduling to Overlap Computation and Communication

You've captured the core problem here: after obtaining an optimal "placement" via **hypergraph partitioning** in section 4.2, **it is highly inefficient** if each device simply **"sequentially executes"** its assigned computation and communication tasks. This approach would cause compute units and network bandwidth to be alternately idle, failing to saturate hardware resources.

Therefore, DCP proposes a **"multi-division execution"** scheduler to solve this:

1.  **Division and Fusing:**
    As you said, the solution first **groups** all computation blocks on a single device into $T$ **"Divisions"**. The benefit of this is that multiple scattered operations within one division can be **"fused"** into a single, more efficient, large kernel, reducing launch overhead.

2.  **Pipeline and Overlap:**
    Most critically, these $T$ divisions are used to construct a **Pipeline**. The core idea is to **"Overlap"** computation and communication: while the device is **"computing"** `Division_i`, it is **"simultaneously"** performing the **"communication"** (prefetching) required for `Division_i+1`.

3.  **Scheduling Goal:**
    The objective of this pipeline, as you pointed out, is to **"maximize the overlap"**. To achieve this, the ideal scenario is to make the **"computation time"** and **"communication time"** of each pipeline stage as **consistent as possible**. If the times are mismatched (e.g., 100ms of computation vs. 10ms of communication), the pipeline will "bubble" or stall, and the benefit of overlapping is lost.

4.  **Implementation (The Algorithm in Listing 3):**
    However, finding the optimal grouping that perfectly balances these times is an NP-complete problem. Therefore, to efficiently achieve this goal of "making the times as consistent as possible," the paper proposes the **greedy heuristic algorithm in Listing 3**. The algorithm is centered around this balancing act:
    * It first calculates the total communication and establishes a **"per-division communication budget"** ($1/T$ of total comm).
    * It **kick-starts** the pipeline by prioritizing "no-communication" blocks.
    * It then greedily fills each stage (prioritizing the device with the lowest computation load) but **defers** any block whose communication cost would **exceed the budget**.
    * In this way, it forces the **"communication time" of most stages to remain at a relatively consistent level**, providing an efficient, approximate solution to your goal of "making computation and communication times as consistent as possible."

#### 1. Block Buffers (Memory Management)
This component solves the memory management problem for the thousands of fine-grained blocks.

* **Problem:** Allocating and managing thousands of small, separate blocks in GPU memory would cause severe **memory fragmentation**.
* **Solution:** The executor pre-allocates a **single, large contiguous buffer** for each data block type (e.g., one buffer for all $Q$ blocks, one for all $K$ blocks, etc.).
* **Addressing:** Blocks are no longer accessed by individual pointers but by their `(type, index)` into the main buffer. This allows kernels (like the modified FlashAttention) to efficiently access data blocks that are **not contiguous in memory**, similar to the `PagedAttention` mechanism. A buffer manager reuses indices of freed blocks to minimize the total memory footprint.

#### 2. DCP Instruction Set (Execution Primitives)
The "Execution Plan" generated by the scheduler is compiled into a sequence of five specialized instructions, which the executor runs sequentially.

* **`Blockwise Attention` (Compute):** A **fused compute kernel** (based on FlashAttention) that executes all attention operations for an entire "Division." It is modified to read its $Q$, $K$, and $V$ inputs from the non-contiguous "Block Buffers."
* **`Blockwise Reduction` (Compute):** A **fused compute kernel** (implemented in Triton) that performs the reduction step, summing all partial outputs into the final $O$ data block.
* **`Blockwise Copy` (Utility):** A fused kernel for efficient, intra-device memory copies, used for managing the block buffers.
* **`Comm. Launch` (Communication):** An **asynchronous** instruction (using PyTorch/NCCL) that **initiates** a P2P data transfer (fetch or send). It returns immediately, allowing the executor to proceed to the next instruction without waiting.
* **`Comm. Wait` (Synchronization):** A **synchronous** instruction that stalls the GPU stream, forcing it to **wait** until a previously launched communication has fully completed.

The core pipeline from Part 3 is realized by the sequential execution of this plan. The executor can run a `Blockwise Attention` instruction (computation) *after* a `Comm. Launch` instruction (communication for a *future* division), thus naturally achieving the desired **overlap** of computation and communication.

Evaluation:

<img width="1067" height="510" alt="Image" src="https://github.com/user-attachments/assets/717d42cc-8465-4d0b-9c83-589c6c1a31f5" />

<img width="573" height="834" alt="Image" src="https://github.com/user-attachments/assets/a459e6e5-9d97-4277-97ee-9165f5b4f20c" />

<img width="528" height="754" alt="Image" src="https://github.com/user-attachments/assets/e61ded11-0e96-4906-be72-e2b662cd889d" />
